examples/blake2f.rs

```diff
@@ -0,0 +1,1427 @@
+    (x, y): (u64, u64),
+    (va_bit_vec, vb_bit_vec): (&mut Vec<Vec<F>>, &mut Vec<Vec<F>>),
+    (vc_bit_vec, vd_bit_vec): (&mut Vec<Vec<F>>, &mut Vec<Vec<F>>),
+    (v_xor_d_bit_vec, v_xor_b_bit_vec): (&mut Vec<Vec<F>>, &mut Vec<Vec<F>>),
+    (b_bit_vec, b_3bits_vec): (&mut Vec<F>, &mut Vec<F>),
+) {
+    va_bit_vec.push(split_value_4bits(
+        v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + x as u128,
+        BITS_NUMBER + 1,
+    ));
+    v1_vec_values[a] = (v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + x as u128) as u64;
+
+    vd_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+    v1_vec_values[d] = ((v1_vec_values[d] ^ v1_vec_values[a]) >> R1)
+        ^ (v1_vec_values[d] ^ v1_vec_values[a]) << (64 - R1);
+    v_xor_d_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+
+    vc_bit_vec.push(split_value_4bits(
+        v1_vec_values[c] as u128 + v1_vec_values[d] as u128,
+        BITS_NUMBER + 1,
+    ));
+    v1_vec_values[c] = (v1_vec_values[c] as u128 + v1_vec_values[d] as u128) as u64;
+
+    vb_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+    v1_vec_values[b] = ((v1_vec_values[b] ^ v1_vec_values[c]) >> R2)
+        ^ (v1_vec_values[b] ^ v1_vec_values[c]) << (64 - R2);
+    v_xor_b_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+
+    va_bit_vec.push(split_value_4bits(
+        v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + y as u128,
+        BITS_NUMBER + 1,
+    ));
+    v2_vec_values[a] = (v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + y as u128) as u64;
+
+    vd_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+    v2_vec_values[d] = ((v1_vec_values[d] ^ v2_vec_values[a]) >> R3)
+        ^ (v1_vec_values[d] ^ v2_vec_values[a]) << (64 - R3);
+    v_xor_d_bit_vec.push(split_value_4bits(v2_vec_values[d] as u128, BITS_NUMBER));
+
+    vc_bit_vec.push(split_value_4bits(
+        v1_vec_values[c] as u128 + v2_vec_values[d] as u128,
+        BITS_NUMBER + 1,
+    ));
+    v2_vec_values[c] = (v1_vec_values[c] as u128 + v2_vec_values[d] as u128) as u64;
+
+    vb_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+    v2_vec_values[b] = ((v1_vec_values[b] ^ v2_vec_values[c]) >> R4)
+        ^ (v1_vec_values[b] ^ v2_vec_values[c]) << (64 - R4);
+    v_xor_b_bit_vec.push(split_value_4bits(
+        (v1_vec_values[b] ^ v2_vec_values[c]) as u128,
+        BITS_NUMBER,
+    ));
+    let bits = (v1_vec_values[b] ^ v2_vec_values[c]) / 2u64.pow(60);
+    b_bit_vec.push(F::from(bits / 8));
+    b_3bits_vec.push(F::from(bits % 8))
+}
+
+fn g_setup<F: PrimeField + Hash>(
+    ctx: &mut StepTypeSetupContext<'_, F>,
+    params: CircuitParams,
+    q_params: GStepParams<F>,
+    (a, b, c, d): (usize, usize, usize, usize),
+    (move1, move2): (u64, u64),
+    s: usize,
+    flag: bool,
+) {
+    // check v[a] = (v[a] + v[b] + x) mod 2^64
```

Comment:
@10to4 it seems that here you are doing:

```
 v[a] := (v[a] + v[b] + x) mod 2**w
v[d] := (v[d] ^ v[a]) >>> R1
v[c] := (v[c] + v[d])     mod 2**w
 v[b] := (v[b] ^ v[c]) >>> R2
 ```
 
But not:

```
v[a] := (v[a] + v[b] + y) mod 2**w
v[d] := (v[d] ^ v[a]) >>> R3
 v[c] := (v[c] + v[d])     mod 2**w
 v[b] := (v[b] ^ v[c]) >>> R4
 ```
 
 Or am I missing something?

Reply:
This two parts are similar. In order to reduce repetitive code, both the parts are implemented by `g_step` function with different parameters. In the spec, the `function G` is called 8 times, so we need to called the `g_step` function 16 times.